marlowe/src/Language/Marlowe/Scripts.hs

```diff
@@ -171,6 +179,122 @@ mkValidator :: MarloweParams -> Scripts.ValidatorType MarloweStateMachine
+                        totalPayouts = foldMap snd payoutsByParty
+                        finalBalance = inputBalance + totalIncome - totalPayouts
+                        outConstrs = OutputConstraint
+                                    { ocDatum = marloweData
+                                    , ocValue = finalBalance
+                                    }
+                        in checkOwnOutputConstraint ctx outConstrs
+            preconditionsOk && inputsOk && payoutsOk && checkContinuation
+        Error TEAmbiguousSlotIntervalError -> traceError "E1"
+        Error TEApplyNoMatchError -> traceError "E2"
+        Error (TEIntervalError (InvalidInterval _)) -> traceError "E3"
+        Error (TEIntervalError (IntervalInPastError _ _)) -> traceError "E4"
+        Error TEUselessTransaction -> traceError "E5"
+
+  where
+    checkScriptOutput addr hsh value TxOut{txOutAddress, txOutValue, txOutDatumHash=Just svh} =
+                    txOutValue == value && hsh == Just svh && txOutAddress == addr
+    checkScriptOutput _ _ _ _ = False
+
+    allOutputs :: [TxOut]
+    allOutputs = txInfoOutputs scriptContextTxInfo
+
+    validateInputs :: [Input] -> Bool
+    validateInputs inputs = all validateInputWitness inputs
+      where
+        validateInputWitness :: Input -> Bool
+        validateInputWitness input =
+            case input of
+                IDeposit _ party _ _         -> validatePartyWitness party
+                IChoice (ChoiceId _ party) _ -> validatePartyWitness party
+                INotify                      -> True
+          where
+            validatePartyWitness (PK pk)     = traceIfFalse "S" $ scriptContextTxInfo `txSignedBy` pk
+            validatePartyWitness (Role role) = traceIfFalse "T" -- "Spent value not OK"
+                                               $ Val.singleton rolesCurrency role 1 `Val.leq` valueSpent scriptContextTxInfo
+
+    collectDeposits :: Input -> Val.Value
+    collectDeposits (IDeposit _ _ (Token cur tok) amount) = Val.singleton cur tok amount
+    collectDeposits _                                     = zero
+
+    payoutByParty :: Payment -> AssocMap.Map Party Val.Value
+    payoutByParty (Payment _ (Party party) money) = AssocMap.singleton party money
+    payoutByParty (Payment _ (Account _) _)       = AssocMap.empty
+
+    payoutConstraints :: [(Party, Val.Value)] -> Bool
+    payoutConstraints payoutsByParty = all payoutToTxOut payoutsByParty
+      where
+        payoutToTxOut (party, value) = case party of
+            PK pk  -> traceIfFalse "P" $ value `Val.leq` valuePaidTo scriptContextTxInfo pk
+            Role role -> let
+                dataValue = Datum $ PlutusTx.toBuiltinData role
+                hsh = findDatumHash dataValue scriptContextTxInfo
+                addr = Ledger.scriptHashAddress rolePayoutValidatorHash
+                in traceIfFalse "R" $ any (checkScriptOutput addr hsh value) allOutputs
+
```

Comment:
Why do payments to public keys test value paid using an inequality, but payments to roles use equality?

My understanding of `payoutConstraints` is
1. When paying a public key, `payout specified by contract` `<=` `actual payout in transaction`.
2. When paying a role, `payout specified by contract` `==` `payout of at least one output`.

In principle this test would permit paying more than the contract permits to a party. For a public key, the `<=` logic permits this. For a role, there could be several outputs totaling more than the specified payout so long as one output contains the exact payout value. So this constraint reads "every party gets paid at least what the contract specifies"? In practice, I presume that the logic of particular contracts might constrain against excess payments?